### ISSUE TITLE
Draw extension lines for angle constraints between line segments

### ISSUE THREAD

**Issue by whitequark** [whitequark/solvespace#95]
Pretty confusing now.

**Comment by Evil-Spirit** [comment]
I hate to do such things without operators overloading.
Why we should use ugly code instead of clean and simple one?
Simple to write, simple to maintain, simple to understand by anyone.

**Comment by whitequark** [comment]
The upstream doesn't want operator overloading. So there won't be any.

**Comment by Evil-Spirit** [comment]
![angle](https://cloud.githubusercontent.com/assets/1980537/14403141/0fe3de16-fe6e-11e5-8525-804ab8d3cb1a.png)

**Comment by Evil-Spirit** [comment]
What about appearance like this?

**Comment by whitequark** [comment]
ISO GPS should say something about extension lines...

**Comment by Evil-Spirit** [comment]
This is artificial restriction not based on the reality, we should change his mind.

**Comment by whitequark** [comment]
I don't really see any difference between the styles myself, and it's a lot of pointless churn, so I don't think it's a good idea either.

**Comment by Evil-Spirit** [comment]
Math can be simple at moment while you are in a context, when you lost it, all the + and->Minus() becomes obvious.

**Comment by Evil-Spirit** [comment]
Call for @jwesthues

**Comment by jwesthues** [comment]
Many projects use operator overloading, but many other projects don't. Some popular languages don't support it at all. We could find examples of serious mathematical software in almost any conceivable style. I weakly like that the current style immediately distinguishes between operations on vector or symbolic types and operations on numerical scalar types, and strongly dislike churn.

Are you sure you've done everything possible to make the current style understandable to you? (Indent appropriately? Break long expressions into smaller ones when necessary? Use an editor that matches parentheses well? Do your math on the whiteboard first instead of in the code?)

Returning from that aside, the feature looks nice. Is the arrowhead at 180 degrees a bug, though?

**Comment by Evil-Spirit** [comment]
Some languages (like Java) has no operator overloading. For many people it is the reason why they don't use it for 3d application development. 
Vector algebra introduces vector operations and sybols for it, so why don't use it? why you need to see the vector and the scalars as different operations, this is just the same in abstraction.
There is no any charn. I think every graphics programmer use it while the creating shaders, and there is no any problems to understand where is the vector and where is the scalar. Expressions  are simple and intuitive, looks like everybody except.

  I am sure, while the using operators overloading it is no reason to indent, break long lines, match paretheses, break into small parts, etc. It' just the expression itself. Just math what we used to write, what we used to read and understand.

Yep, 180deg some buggy., Thanks god this is not yet ->plus->minus corrupted and i can found some reason for it :)

**Comment by whitequark** [comment]
The reason people don't use Java for graphics programming is the lack of value types, leading to massive GC churn on trivial operations, not lack of operator overloading.

Churn is the very large amount of code change required to change the style. You have to rewrite pretty much all math.

**Comment by Evil-Spirit** [comment]
And this too.
I've done it once, there is no reason why i can't do it again.

**Comment by whitequark** [comment]
The question is not whether you can, but whether you should. Rewriting code breaks git history and introduces bugs.

**Comment by whitequark** [comment]
@Evil-Spirit Extension lines look nice indeed, let's go ahead with this.

**Comment by Evil-Spirit** [comment]
![angle](https://cloud.githubusercontent.com/assets/1980537/14430616/1783aeaa-0024-11e6-9a01-b4d44ee27d09.png)

**Comment by whitequark** [comment]
Excellent!

**Comment by whitequark** [comment]
Problem 1: extremely large extension lines with small angles like this:
![2045](https://cloud.githubusercontent.com/assets/54771/14436172/f900094c-000a-11e6-842b-b6490e5c2e6d.png)
Problem 2: moving the label around with mouse is very unstable in this case:
![2046](https://cloud.githubusercontent.com/assets/54771/14436187/127ca38a-000b-11e6-8ca3-d2d7f0ae5d36.png)
Problem 3: text intersecting with the lines:
![2047](https://cloud.githubusercontent.com/assets/54771/14436233/49df1f38-000b-11e6-9775-16e341890301.png)

**Comment by whitequark** [comment]
These problems aren't introduced *with* the extension lines patch but we should address them anyway.

**Comment by Evil-Spirit** [comment]
I also suggest to use %g format specifier to avoid pointless ".00"

**Comment by Evil-Spirit** [comment]
1. What we should do in this case?

**Comment by whitequark** [comment]
Hm, good question. I actually don't know. Librecad has the same problem...

**Comment by whitequark** [comment]
Looks like even AutoCAD doesn't have a good answer to 1)

**Comment by whitequark** [comment]
@Evil-Spirit Ah I finally realized what's wrong here. We can solve problems 2 and 3 by making the label for the angle constraint interact with the line in the same way as for the distance constraints:
![2060](https://cloud.githubusercontent.com/assets/54771/14475001/5044cf46-00ef-11e6-8d74-d5bb91f9ddab.png)

**Comment by Evil-Spirit** [comment]
@whitequark, There will be no problems with image you have provided, but for small arcs it will be very bad appearance. OffsetCalculation (using EllipticalInterpolation) is performed in wrong way.
We should do something like this http://pyx.sourceforge.net/manual/box.html

**Comment by whitequark** [comment]
@Evil-Spirit Yes. We already have that problem with distance constraints though:
![2062](https://cloud.githubusercontent.com/assets/54771/14480148/a3550216-0115-11e6-985c-f62c0783e25f.png)

I think what we should do is use the current format (dimension in middle of a line) for both angle and distance constraints, and, separately, detect when the dimension line is too small and take the label out of it. Then we don't even need your fancy algorithm, because you can assume that the line is nearly straight.

**Comment by Evil-Spirit** [comment]
![2016-04-13 10 42 27](https://cloud.githubusercontent.com/assets/1980537/14483048/9530b546-0164-11e6-874a-dac70dcfe670.png)

**Comment by Evil-Spirit** [comment]
Problem 1:Don't know how to fix
Problem 2:Wontfix; just fix it in next iteration when we will introduce constraint offset basis.
Problem 3:Solved

**Comment by whitequark** [comment]
Great!

**Comment by whitequark** [comment]
@Evil-Spirit Can you also make the arrows invert and the line extend, like with distance constraints?
![2067](https://cloud.githubusercontent.com/assets/54771/14507481/4c1ece8a-01b2-11e6-8387-fe4a391d0f0f.png)

**Comment by Evil-Spirit** [comment]
@whitequark OK, forget about it.

**Comment by whitequark** [comment]
@Evil-Spirit can you please finish this first? I'd like to finish this merge and make a nightly.

**Comment by Evil-Spirit** [comment]
Ok, will do it now.

**Comment by Evil-Spirit** [comment]
![2016-04-15 17 55 51](https://cloud.githubusercontent.com/assets/1980537/14560783/acb17f86-0333-11e6-80a9-c95137f34baa.png)

**Comment by whitequark** [comment]
@Evil-Spirit Good but that doesn't include the arrow inversion (like on distance constraints).

**Comment by Evil-Spirit** [comment]
@whitequark, i don't know, but this can cause confusion. I dont found any samples about it.

**Comment by Evil-Spirit** [comment]
@whitequark When we have linear dimension - we have straight line which is not closed-loop. And we can treat it correctly. When we have angular dimension, we have an arc which is loops in circle, and there is ambigous: we can treat it an aplha angle or 360-alpha.

**Comment by whitequark** [comment]
@Evil-Spirit This is specified in ISO--e.g. see page 8 and especially page 38 of http://engineering.pages.tcnj.edu/files/2012/02/dimensioning_and_tolerancing.pdf. We really need to get a copy of ISO 128-22 somewhere...

**Comment by Evil-Spirit** [comment]
@whitequark, I saw this document, but found only 8th page. Page 8th case can be applicable to the near 0 or 180 degrees cases. For page 38 we shoud consider what degree value is inside, but we have outside case and I have no opinion what we should do with.

**Comment by whitequark** [comment]
@Evil-Spirit The direction of the arrow does not say anything about the angle--it's a part of the leader line from the label to the dimension being specified. I.e. think of the indicated angle as one separate thing and a label with a leader and an arrow pointing from it as another thing.

There isn't any ambiguity because angles over 180° aren't conventionally used.

See also https://ia801408.us.archive.org/18/items/BasicEngineeringDrawing/RhodesCook-BasicEngineeringDrawing.pdf for future reference starting at page 70--though it doesn't say anything new on this specific issue.

**Comment by whitequark** [comment]
Actually, even if angles over 180° would be used there wouldn't be any real ambiguity because it is quite obvious when an angle is larger than half a circle.

**Comment by Evil-Spirit** [comment]
> There isn't any ambiguity because angles over 180° aren't conventionally used.

Not used YET. We have to introduce it in future (for measuring arc angle which is actually 0-360).

**Comment by whitequark** [comment]
Do we need that? It's very rarely used and you can just do some construction:
![2086](https://cloud.githubusercontent.com/assets/54771/14579987/5ed3870c-03ae-11e6-9afb-e15381167de8.png)

Maybe we should allow adding an angle constraint between three points, but even that seems extraneous.

**Comment by whitequark** [comment]
Also, let's say we do want to specify an angle over 180° degree. Then the dimension line should look something like this (pardon my crude sketching):
![2087](https://cloud.githubusercontent.com/assets/54771/14580038/379a57b8-03b0-11e6-8266-c60db344d404.png)

So there's also no ambiguity.

**Comment by Evil-Spirit** [comment]
> Do we need that? It's very rarely used and you can just do some construction:

So, try to constrain 100 arcs by involving additional segments. There will be churn.

![2016-04-16 15 21 45](https://cloud.githubusercontent.com/assets/1980537/14580192/38896e5c-03e7-11e6-9022-3a2f0437b8a3.png)

**Comment by Evil-Spirit** [comment]
https://youtu.be/b41zpAzhH_M?list=PLqmaFsEpxcUXzvoLdZ7QtpKb_iInqo1pT

**Comment by whitequark** [comment]
The video contains a very artificial example. People don't actually use parametric modeling like that, it's a job more suited to Blender or something. And overall, angles over 180° are extremely rare; in the sketch you posted above, the dimension used to constrain the inner flat would be the distance to center, or perhaps distance to opposite side of the arc.

**Comment by Evil-Spirit** [comment]
Anyway, we have to do the best way if we know how. The degrees > 180 are useful, angle for arcs are useful (without any additional sketching), and arc lengths are useful (there will be the same visualization as angle).

**Comment by whitequark** [comment]
I don't think they are useful, I don't see any compelling use cases, and as such implementing them is a waste of time. And regardless of that there isn't any ambiguity anyway as I've shown above.

**Comment by whitequark** [comment]
> there isn't any ambiguity

Specifically, the angle being constrained/measured is indicated by a reference arc across that angle. You don't get any less ambiguous. The label may occlude it entirely but it's still unambiguous because there is no reference arc on the *other* side.

**Comment by Evil-Spirit** [comment]
There is no reason why we should'nt go thorugh the footpath I already pass. It's time already wasted, so we can have it in much less time. The sooner or later. The small things like this doing all the imression. Devil in the details. etc.

**Comment by whitequark** [comment]
We have to add that to the file format, maintain it, fix bugs in it, etc. I would rather not do all that for a useless feature. If a feature has no compelling use case, I don't care how easy it is to add it, I won't add it, because features used by no one stay alive for years anyway and extract their maintenance cost.

**Comment by Evil-Spirit** [comment]
@whitequark , 
@jwesthues ,
there is no specific; just gui work. anyway we MUST do atan2 angle in order to support 	chirality(enantiomorphism) of the sketches. Just look at ugly workaround to hide acos(dot) angle equation problems:
```
                // avoid false detection of inconsistent systems by gaining
                // up as the difference in dot products gets small at small
                // angles; doubles still have plenty of precision, only
                // problem is that rank test
                Expr *mult = Expr::From(arc > 0.99 ? 0.01/(1.00001 - arc) : 1);
```

**Comment by whitequark** [comment]
Yes, atan2 will need to be done, but that happens after refactoring and once we have migrated to a new file format, as discussed before.

Until then, would you please focus on the particular issue and fix the arrows? It's the last remaining part right now...

**Comment by Evil-Spirit** [comment]
![2016-04-16 17 59 56](https://cloud.githubusercontent.com/assets/1980537/14580836/2cddd03c-03fd-11e6-86ff-5584f4d22d16.png)

**Comment by whitequark** [comment]
Excellent, thanks! Fixed in a4a353f01.
